src/services/blocks/trace/Trace.ts

```diff
@@ -0,0 +1,666 @@
+		// are this program has some incorrect assumptions and this error should
+		// be reported to the maintainers.
+		for (const [i, extIdx] of extrinsicIndex.entries()) {
+			if (i > 0 && !extrinsicIndex[i]?.eq(extIdx)) {
+				throw new InternalServerError(
+					'Expect extrinsic to only be applied at a single index.'
+				);
+			}
+		}
+	}
+
+	return extrinsicIndex[0];
+}
+
+/**
+ * Class for processing traces from the `state_traceBlock` RPC endpoint.
+ *
+ * Assumptions:
+ * - Spans do not start in sorted order.
+ * - Events start in sorted order.
+ *
+ * For a conceptual overview on traces from block excution and `Operation` construction
+ * consult [this diagram.](https://docs.google.com/drawings/d/1vZoJo9jaXlz0LmrdTOgHck9_1LsfuQPRmTr-5g1tOis/edit?usp=sharing)
+ */
+export class Trace {
+	/**
+	 * Known storage keys.
+	 */
+	private keyNames;
+	constructor(
+		api: ApiPromise,
+		/**
+		 * `state_traceBlock` RPC response `result`.
+		 */
+		private traceBlock: BlockTrace,
+		/**
+		 * Type registry corresponding to the exact runtime the traces are from.
+		 */
+		private registry: Registry
+	) {
+		if (!traceBlock.spans.length) {
+			throw new InternalServerError(
+				'No spans found. This runtime is likely not supported with tracing.'
+			);
+		}
+
+		this.keyNames = getKeyNames(api);
+	}
+
+	/**
+	 * Derive the action groups and operations from the block trace.
+	 *
+	 * @returns the action groups and opertions
+	 */
+	actionsAndOps(): { actions: ActionGroup[]; operations: Operation[] } {
+		const { spansById, parsedSpans: spans, executeBlockSpanId } = parseSpans(
+			this.traceBlock.spans
+		);
+		const { eventsByParentId, extrinsicIndexBySpanId } = this.parseEvents(
+			spansById
+		);
+
+		const operations: Operation[] = [];
+		const actions: ActionGroup[] = [];
+		// Create a list of action groups (`actions`) and a list of all `Operations`.
+		for (const primary of spans) {
+			if (!(primary.parentId === executeBlockSpanId)) {
+				// Only use primary spans (spans where parentId === execute_block). All other
+				// spans are either the executeBlockSpan or descendant of a primary span.
+				continue;
+			}
+
+			// Gather the secondary spans and their events. Events are useful for
+			// creating Operations. Spans are useful for attributing events to a
+			// hook or extrinsic.
+			// Keep in mind events are storage Get/Put events. The spans are used to
+			// attribute events to some action in the block execution pipeline.
+			const secondarySpansIds = findDescendants(primary.id, spansById);
+			const secondarySpansEvents: PEvent[] = [];
+			const secondarySpans: SpanWithChildren[] = [];
+			secondarySpansIds.forEach((id) => {
+				const events = eventsByParentId.get(id);
+				const span = spansById.get(id);
+
+				events && secondarySpansEvents.push(...events);
+				span && secondarySpans.push(span);
+			});
+
+			let phase: PhaseOther, extrinsicIndex: BN | undefined;
```

Comment:
```suggestion
			let phase?: PhaseOther, extrinsicIndex: BN;
```

Reply:
Only works for interfaces and function params afaict.